tests/org.eclipse.emf.test.core/src/org/eclipse/emf/test/core/ecore/OSGiDelegatEPackageRegistryTest.java

```diff
@@ -0,0 +1,205 @@
+    {
+      serviceTracker.close();
+    }
+
+    for (ServiceRegistration<?> serviceRegistration : registrations)
+    {
+      try
+      {
+        serviceRegistration.unregister();
+      }
+      catch (IllegalStateException exception)
+      {
+      }
+    }
+  }
+
+  private EPackage.Registry newInstance()
+  {
+    try
+    {
+      EPackage.Registry newInstance = OSGI_DELEGATE_PACKAGE_REGISTRY_CONSTRUCTOR.newInstance(BUNDLE_CONTEXT);
+      ServiceTracker<?, ?> serviceTracker = new ServiceTracker<>(BUNDLE_CONTEXT, FILTER, (ServiceTrackerCustomizer<?, ?>)newInstance);
+      trackers.add(serviceTracker);
+      serviceTracker.open();
+      return newInstance;
+    }
+    catch (Exception exception)
+    {
+      throw new RuntimeException(exception);
+    }
+  }
+
+  private Dictionary<String, Object> newDictionary(int serviceRanking)
+  {
+    Map<String, Object> properties = new LinkedHashMap<>();
+    properties.put("emf.default.epackage.registry", Boolean.TRUE);
+    properties.put(Constants.SERVICE_RANKING, serviceRanking);
+    return FrameworkUtil.asDictionary(properties);
+  }
+
+  private ServiceRegistration<Registry> newServiceRegistration(EPackage.Registry registry, int serviceRanking)
+  {
+    ServiceRegistration<Registry> serviceRegistration = BUNDLE_CONTEXT.registerService(Registry.class, registry, newDictionary(serviceRanking));
+    registrations.add(serviceRegistration);
+    return serviceRegistration;
+  }
+
+  private void compare(EPackage.Registry registry1, EPackage.Registry registry2)
+  {
+    // Ensure that all the methods of the implementation are tested a part of testing for equality.
+    assertEquals(registry1.isEmpty(), registry2.isEmpty());
+    assertEquals(registry1.size(), registry2.size());
+    assertEquals(registry1.entrySet(), registry2.entrySet());
+    assertEquals(registry1.keySet(), registry2.keySet());
+    assertEquals(registry1.values().size(), registry2.values().size());
+    assertEquals(new HashSet<>(registry1.values()), new HashSet<>(registry2.values()));
+    assertEquals(registry1, registry2);
+  }
+
+  private void validate(EPackage.Registry osgiEPackageRegistry)
+  {
+    // Ensure that all the methods of the implementation are tested.
+    assertTrue(osgiEPackageRegistry.containsKey("foo"));
+    assertTrue(osgiEPackageRegistry.containsValue(EcorePackage.eINSTANCE));
+    assertEquals(EcorePackage.eINSTANCE, osgiEPackageRegistry.get("foo"));
+    assertEquals(EcorePackage.eINSTANCE, osgiEPackageRegistry.remove("foo"));
+    osgiEPackageRegistry.putAll(Collections.singletonMap("foo", EcorePackage.eINSTANCE));
+    osgiEPackageRegistry.clear();
+    assertTrue(osgiEPackageRegistry.isEmpty());
+    osgiEPackageRegistry.putAll(Collections.singletonMap("foo", EcorePackage.eINSTANCE));
+    assertEquals(EcorePackage.eINSTANCE, osgiEPackageRegistry.getEPackage("foo"));
+    assertEquals(EcoreFactory.eINSTANCE, osgiEPackageRegistry.getEFactory("foo"));
+  }
```

Comment:
This helps provide coverage for all the method of the implementation both while there is a delegate and while there isn't.